backpack/extensions/module_extension.py

```diff
@@ -12,98 +24,156 @@ class ModuleExtension:
+    def backpropagate(
+        self,
+        extension: BackpropExtension,
+        module: Module,
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        bpQuantities: Any,
+    ) -> Any:
+        """Backpropagation of additional information through the graph.
+
+        Args:
+            extension: Instance of the extension currently running
+            module: Instance of the extended module
+            g_inp: Gradient of the loss w.r.t. the inputs
+            g_out: Gradient of the loss w.r.t. the output
+            bpQuantities: Quantities backpropagated w.r.t. the output
 
         Returns
-        -------
-        bpQuantities:
             Quantities backpropagated w.r.t. the input
         """
         warnings.warn("Backpropagate has not been overwritten")
 
-    def apply(self, ext, module, g_inp, g_out):
-        """
+    def apply(
+        self,
+        extension: BackpropExtension,
+        module: Module,
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+    ) -> None:
+        """Apply all actions required by the extension.
+
         Fetch backpropagated quantities from module output, apply backpropagation
         rule, and attach the result to module input(s).
-        """
-        inp = module.input0
-        out = module.output
 
-        bpQuantities = self.__backproped_quantities(ext, out)
+        Args:
+            extension: current backpropagation extension
+            module: current module
+            g_inp: input gradients
+            g_out: output gradients
+
+        Raises:
+            AssertionError: if there is no saved quantity although extension expects one
+        """
+        bp_quantity = self.__get_backproped_quantity(extension, module.output)
+        if (
+            extension.expects_backpropagation_quantities() is True
+            and bp_quantity is None
+            and is_loss(module) is False
+            and TORCH_VERSION_HIGHER_THAN_1_9_0
+        ):
+            raise AssertionError(
+                "BackPACK extension expects a backpropagation quantity but it is None. "
+                f"Module: {module}, Extension: {extension}."
+            )
 
         for param in self.__params:
             if self.__param_exists_and_requires_grad(module, param):
                 extFunc = getattr(self, param)
-                extValue = extFunc(ext, module, g_inp, g_out, bpQuantities)
-                self.__save(extValue, ext, module, param)
+                extValue = extFunc(extension, module, g_inp, g_out, bp_quantity)
+                self.__save_value_on_parameter(extValue, extension, module, param)
+
+        if extension.expects_backpropagation_quantities():
+            bp_quantity = self.backpropagate(
+                extension, module, g_inp, g_out, bp_quantity
+            )
+            self.__save_backprop_quantity(
+                extension,
+                module.input0,
+                bp_quantity,
+            )
```

Comment:
`if not keep_graph: # delete backpropagated quantity on output`

Reply:
Realized, but at the beginning of the `__call__` function instead of at the end.

Reply:
Yes.